zebra-chain/src/work/equihash.rs

```diff
@@ -60,11 +65,120 @@ impl Solution {
         Ok(())
     }
 
-    #[cfg(feature = "getblocktemplate-rpcs")]
+    /// Returns a [`Solution`] containing the bytes from `solution`.
+    /// Returns an error if `solution` is the wrong length.
+    pub fn from_bytes(solution: &[u8]) -> Result<Self, SerializationError> {
+        if solution.len() != SOLUTION_SIZE {
+            return Err(SerializationError::Parse(
+                "incorrect equihash solution size",
+            ));
+        }
+
+        let mut bytes = [0; SOLUTION_SIZE];
+        // Won't panic, because we just checked the length.
+        bytes.copy_from_slice(solution);
+
+        Ok(Self(bytes))
+    }
+
     /// Returns a [`Solution`] of `[0; SOLUTION_SIZE]` to be used in block proposals.
+    #[cfg(feature = "getblocktemplate-rpcs")]
     pub fn for_proposal() -> Self {
         Self([0; SOLUTION_SIZE])
     }
+
+    /// Mines and returns one or more [`Solution`]s based on a template `header`.
+    /// The returned header contains a valid `nonce` and `solution`.
+    ///
+    /// If `cancel_fn()` returns an error, returns early with `Err(SolverCancelled)`.
+    ///
+    /// The `nonce` in the header template is taken as the starting nonce. If you are running multiple
+    /// solvers at the same time, start them with different nonces.
+    /// The `solution` in the header template is ignored.
+    ///
+    /// This method is CPU and memory-intensive. It uses 144 MB of RAM and one CPU core while running.
+    /// It can run for minutes or hours if the network difficulty is high.
+    #[cfg(feature = "internal-miner")]
+    #[allow(clippy::unwrap_in_result)]
+    pub fn solve<F>(mut header: Header, mut cancel_fn: F) -> Result<Header, SolverCancelled>
+    where
+        F: FnMut() -> Result<(), SolverCancelled>,
+    {
+        use crate::shutdown::is_shutting_down;
+
+        let mut input = Vec::new();
+        header
+            .zcash_serialize(&mut input)
+            .expect("serialization into a vec can't fail");
+        let input = &input[0..Solution::INPUT_LENGTH];
+
+        while !is_shutting_down() {
+            // Don't run the solver if we'd just cancel it anyway.
+            cancel_fn()?;
+
+            let solutions = equihash_solver::tromp::solve_200_9_compressed(input, || {
```

Comment:
This is incorrect, because INPUT_LENGTH is shorter than the length of the serialized header. (It excludes the nonce, solution length, and solution.)

I'll add a comment instead.

Reply:
Documented in 6e6bc01c19